File: .tmp_analyze_si_candidates.js
const fs = require('fs');
const path = require('path');

// Load audit results
const results = JSON.parse(fs.readFileSync('.tmp_retrieval_results.json', 'utf8'));

console.log('='.repeat(120));
console.log('SI CHUNK CANDIDATE ANALYSIS');
console.log('='.repeat(120));

for (let queryIdx = 0; queryIdx < results.length; queryIdx++) {
  const r = results[queryIdx];
  const question = r.question || 'N/A';
  const intent = r.intent || 'N/A';
  const userIntent = r.userIntent || 'N/A';
  
  console.log(`\n${'='.repeat(120)}`);
  console.log(`QUERY ${queryIdx + 1}: ${question}`);
  console.log(`Intent: ${intent} | UserIntent: ${userIntent}`);
  console.log(`${'='.repeat(120)}`);
  
  const top20 = r.top20 || [];
  const relevantIds = new Set((r.relevantIds || []).map(item => item.id));
  
  // Extract SI candidates from top 20
  const siCandidates = [];
  
  for (let rank = 0; rank < top20.length; rank++) {
    const scoredItem = top20[rank];
    const item = scoredItem.item || {};
    const itemId = item.id || 'unknown';
    const filename = item.filename || item.trainingId || 'unknown';
    const program = item.program || item.programName || 'N/A';
    const docCategory = scoredItem.docCategory || item.docCategory || 'N/A';
    
    // Skip non-SI programs
    if (String(program).toUpperCase() !== 'SI') {
      continue;
    }
    
    // Skip Double Degree
    if (filename.toLowerCase().includes('double') || String(item.chunk || '').toLowerCase().includes('double')) {
      continue;
    }
    
    // This is a valid SI candidate
    const semanticScore = scoredItem.score || 0;
    const compositeScore = scoredItem.compositeScore || 0;
    const finalScore = scoredItem.finalScore || 0;
    
    // Check if passed filterRelevantChunks
    const passedFilter = relevantIds.has(itemId);
    
    siCandidates.push({
      rank: rank + 1,
      id: itemId,
      filename: filename,
      category: docCategory,
      semanticScore: semanticScore,
      compositeScore: compositeScore,
      finalScore: finalScore,
      passedFilter: passedFilter,
      chunkPreview: String(item.chunk || '').substring(0, 150)
    });
  }
  
  if (siCandidates.length === 0) {
    console.log('\n[NONE] NO SI CANDIDATES found in top 20 (excluding Double Degree)');
    console.log('\nTop 5 program distribution:');
    for (let i = 0; i < Math.min(5, top20.length); i++) {
      const scoredItem = top20[i];
      const item = scoredItem.item || {};
      const filename = item.filename || item.trainingId || 'unknown';
      const program = item.program || item.programName || 'N/A';
      const docCategory = scoredItem.docCategory || item.docCategory || 'N/A';
      const compositeScore = scoredItem.compositeScore || 0;
      console.log(`  ${i+1}. ${filename.substring(0, 50).padEnd(50)} | program=${program} | category=${docCategory} | composite=${compositeScore.toFixed(4)}`);
    }
  } else {
    console.log(`\n[YES] Found ${siCandidates.length} SI candidates in top 20:`);
    console.log();
    
    // Show best SI candidate
    const bestSI = siCandidates[0];
    console.log('>>> BEST SI CANDIDATE <<<');
    console.log(`  Rank #${bestSI.rank} | [${bestSI.passedFilter ? 'PASSED' : 'REJECTED'}] filterRelevantChunks`);
    console.log(`    ID: ${bestSI.id}`);
    console.log(`    Filename: ${bestSI.filename}`);
    console.log(`    Category: ${bestSI.category}`);
    console.log(`    Semantic Score: ${bestSI.semanticScore.toFixed(4)}`);
    console.log(`    Composite Score: ${bestSI.compositeScore.toFixed(4)}`);
    console.log(`    Chunk Preview: ${bestSI.chunkPreview}`);
    console.log();
    
    // Show all SI candidates summary
    console.log('All SI candidates:');
    for (const cand of siCandidates) {
      const status = cand.passedFilter ? 'PASSED' : 'REJECTED';
      console.log(`  #${cand.rank.toString().padStart(2)} [${status}] composite=${cand.compositeScore.toFixed(4)} | ${cand.filename.substring(0, 40)}`);
    }
  }
  
  // Summary statistics
  console.log(`\nSUMMARY:`);
  console.log(`  - Top 20 count: ${top20.length}`);
  console.log(`  - SI candidates (excluding DD): ${siCandidates.length}`);
  console.log(`  - SI candidates PASSED filter: ${siCandidates.filter(c => c.passedFilter).length}`);
  console.log(`  - SI candidates REJECTED: ${siCandidates.filter(c => !c.passedFilter).length}`);
  console.log(`  - Relevant IDs after filter: ${relevantIds.size}`);
  
  // Show what passed the filter
  if (relevantIds.size > 0) {
    console.log(`\n  [INFO] Chunks that PASSED filterRelevantChunks:`);
    for (const item of (r.relevantIds || [])) {
      console.log(`    - ${item.filename.substring(0, 50).padEnd(50)} | category=${item.docCategory} | composite=${parseFloat(item.compositeScore).toFixed(4)}`);
    }
  }
}

console.log('\n' + '='.repeat(120));
console.log('END OF ANALYSIS');
console.log('='.repeat(120));

File: .tmp_analyze_si_candidates.py
#!/usr/bin/env python3
"""
Analyze SI chunk candidates from top 20 before filtering.
For each query SI benchmark, identify:
1. SI chunks (program=SI, not MI/SK/DD) from top 20
2. Their ranking, semantic score, composite score
3. Whether they passed filterRelevantChunks (in relevantIds)
4. If rejected, which rule likely rejected them
"""

import json
from pathlib import Path

# Load audit results
results_path = Path('.tmp_retrieval_results.json')
results = json.loads(results_path.read_text('utf8'))

print("=" * 120)
print("SI CHUNK CANDIDATE ANALYSIS")
print("=" * 120)

for query_idx, r in enumerate(results):
    question = r.get('question', 'N/A')
    intent = r.get('intent', 'N/A')
    user_intent = r.get('userIntent', 'N/A')
    
    print(f"\n{'='*120}")
    print(f"QUERY {query_idx + 1}: {question}")
    print(f"Intent: {intent} | UserIntent: {user_intent}")
    print(f"{'='*120}")
    
    top20 = r.get('top20', [])
    relevant_ids = set(item['id'] for item in r.get('relevantIds', []))
    
    # Extract SI candidates from top 20
    # SI = program=SI but not (MI or SK or "Double Degree")
    si_candidates = []
    
    for rank, scored_item in enumerate(top20, 1):
        item = scored_item.get('item', {})
        item_id = item.get('id', 'unknown')
        filename = item.get('filename') or item.get('trainingId') or 'unknown'
        program = item.get('program') or item.get('programName') or 'N/A'
        doc_category = scored_item.get('docCategory') or item.get('docCategory') or 'N/A'
        
        # Skip non-SI programs
        if str(program).upper() != 'SI':
            continue
        
        # Skip Double Degree (already tracked separately)
        if 'double' in filename.lower() or 'double' in str(item.get('chunk', '')).lower():
            continue
        
        # This is a valid SI candidate
        semantic_score = scored_item.get('score', 0)
        composite_score = scored_item.get('compositeScore', 0)
        final_score = scored_item.get('finalScore', 0)
        
        # Check if passed filterRelevantChunks
        passed_filter = item_id in relevant_ids
        
        si_candidates.append({
            'rank': rank,
            'id': item_id,
            'filename': filename,
            'category': doc_category,
            'semantic_score': semantic_score,
            'composite_score': composite_score,
            'final_score': final_score,
            'passed_filter': passed_filter,
            'chunk_preview': str(item.get('chunk', ''))[:150]
        })
    
    if not si_candidates:
        print("\n[NONE] NO SI CANDIDATES found in top 20 (excluding Double Degree)")
        print("\nTop 20 program distribution:")
        for rank, scored_item in enumerate(top20[:5], 1):
            item = scored_item.get('item', {})
            filename = item.get('filename') or item.get('trainingId') or 'unknown'
            program = item.get('program') or item.get('programName') or 'N/A'
            doc_category = scored_item.get('docCategory') or item.get('docCategory') or 'N/A'
            composite_score = scored_item.get('compositeScore', 0)
            print(f"  {rank}. {filename[:50]:50} | program={program} | category={doc_category} | composite={composite_score:.4f}")
    else:
        print(f"\n[YES] Found {len(si_candidates)} SI candidates in top 20:")
        print()
        
        for candidate in si_candidates:
            status = "[PASSED] filterRelevantChunks" if candidate['passed_filter'] else "[REJECTED] by filterRelevantChunks"
            print(f"  Rank #{candidate['rank']} | {status}")
            print(f"    ID: {candidate['id']}")
            print(f"    Filename: {candidate['filename']}")
            print(f"    Category: {candidate['category']}")
            print(f"    Semantic Score: {candidate['semantic_score']:.4f}")
            print(f"    Composite Score: {candidate['composite_score']:.4f}")
            print(f"    Final Score: {candidate['final_score']:.4f}")
            print(f"    Chunk Preview: {candidate['chunk_preview']}")
            print()
    
    # Summary statistics
    print(f"\nSUMMARY:")
    print(f"  - Top 20 count: {len(top20)}")
    print(f"  - SI candidates: {len(si_candidates)}")
    print(f"  - SI candidates passed filter: {sum(1 for c in si_candidates if c['passed_filter'])}")
    print(f"  - SI candidates rejected: {sum(1 for c in si_candidates if not c['passed_filter'])}")
    print(f"  - Relevant IDs after filter: {len(relevant_ids)}")
    
    # Show what passed the filter
    if relevant_ids:
        print(f"\n  [INFO] Chunks that PASSED filterRelevantChunks:")
        for item in r.get('relevantIds', []):
            print(f"    - {item['filename'][:50]:50} | category={item['docCategory']} | composite={float(item['compositeScore']):.4f}")

print("\n" + "=" * 120)
print("END OF ANALYSIS")
print("=" * 120)
